### PR TITLE
add python api to get headers only from files

### PR DESCRIPTION
**Pull request by i30817**
As in the bug report except get_parent_sha1 is returning a option instead of a 20 bytes 0 filled vector. Not sure if you want that, or would prefer a exception instead, but i thought python 'None' was better suited to represent 'no parent' than the zero filled 20 bytes; and it still fails if you test with `if header.get_parent_sha1 == other.get_sha1`.

Now i just want a way to extract the sha1 of only each track instead of the total...

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -32,6 +33,91 @@ struct Metadata {
...
+    pub fn has_parent(&self) -> bool
+    {
+        self.inner.has_parent()
+    }
+
+    pub fn len(&self) -> u32 {
+        self.inner.len()
+    }
+
+    pub fn get_sha1(&self) -> &[u8] {
```
Comment (chyyran) [inline src/lib.rs]: Can you rename this to `sha1` to be consistent with the other methods, and have it return `Option<&[u8]>`? That way you can return `Some(h)` for V5, and `None` for non-v5.

File: src/lib.rs [hunk 2]
```diff
@@ -32,6 +33,91 @@ struct Metadata {
...
+        let h = match &self.inner {
+            V5Header(h) => h,
+            _ => unimplemented!(),
+        };
+        &h.sha1
+    }
+
+    pub fn get_parent_sha1(&self) -> Option<&[u8]> {
```
Comment (chyyran) [inline src/lib.rs]: as above, return `None` for non V5 and rename to `parent_sha1`

File: src/lib.rs [hunk 3]
```diff
@@ -32,6 +33,91 @@ struct Metadata {
...
+    {
+        self.inner.unit_count()
+    }
+
+    pub fn has_parent(&self) -> bool
+    {
+        self.inner.has_parent()
+    }
+
+    pub fn len(&self) -> u32 {
```
Comment (chyyran) [inline src/lib.rs]: needs `#[pyo3(name = "__len__")]` so `len(header)` can be called in python

File: src/lib.rs [hunk 4]
```diff
@@ -32,6 +33,91 @@ struct Metadata {
...
+            _ => unimplemented!(),
+        };
+        if self.inner.has_parent() {
+            return Some(&h.parent_sha1)
+        } else {
+            return None
+        }
+    }
+
+    pub fn get_raw_sha1(&self) -> &[u8] {
```
Comment (chyyran) [inline src/lib.rs]: as above, return `Option<&[u8]>`, and rename to `raw_sha1`

File: src/lib.rs [hunk 5]
```diff
@@ -91,11 +177,32 @@ fn chd_open(path: String, parent: Option<String>) -> PyResult<Chd> {
     })
 }
 
+#[pyfunction]
+fn try_read_headers(paths: Vec<String>) -> PyResult<Vec<Header>> {
```
Comment (chyyran) [inline src/lib.rs]: Rename this to `chd_read_header`

I would rather this take in a single string and return a single header, i.e. 

`fn try_read_header(path: String) -> PyResult<Header>`. Then the for loop can be written in Python or as a list comprehension.

```python
files = [...]
headers = [chd_read_header(f) for f in files]
```

Additionally there's no reason to restrict this to only V5 headers, if the V5 exclusive methods return `Option<&[u8]>` (which will map to `None` in Python for non-V5 headers).

**Review by chyyran — CHANGES_REQUESTED** [review]

**Comment by chyyran** [comment]
Looks mostly good, just some small changes needed.

**Comment by chyyran** [comment]
> Now i just want a way to extract the sha1 of only each track instead of the total...

Get the metadata to find the track separations, buffer the track via `hunk` and calculate the SHA1 manually. You can do this in the Python side or in Rust side faster, https://github.com/SnowflakePowered/chd-rs/blob/master/rchdman/src/main.rs#L514 does the SHA1 of the whole thing but if you parse the track metadata you should be able to do it for each track.

File: src/lib.rs (hunk 1 above)
Comment (chyyran) [reply src/lib.rs]: V3 and V4 also have sha1

**Comment by i30817** [comment]
Mmm maybe the return being a list is not the best api

I'm thinking of a environment where there are mixed versions of the header. This code only handles v5 headers, and i'm fine with it, for myself, but imagine someone that places both v5 and v4 and tries to read both in a single call, because they search the filesystem for 'chd' in any way.

Exception. While if the for was external, the exception could be caught. Yeah, i think i'm exposing the single header version of the function instead.

**Comment by chyyran** [comment]
You can add a `version()` method to `Header` that just matches on the header type and returns the version so you can do 

```python
headers = [chd_read_header(f) for f in files]
headers_v5 = [h if h.version() is 5 for h in headers]
```

**Comment by i30817** [comment]
> You can add a `version()` method to `Header` that just matches on the header type and returns the version so you can do
> 
> ```python
> headers = [chd_read_header(f) for f in files]
> headers_v5 = [h if h.version() is 5 for h in headers]
> ```

return type? usize or u32?

Also I had to turn the function len from u32 to usize for the `#[pyo3(name = "__len__")]` property to work.

**Comment by chyyran** [comment]
I believe you have to cast it to `usize` for PyO3

**Comment by i30817** [comment]
It's also different from the method already there in that that method returns a PyResult. Should this one do that too, or is it pointless? 
writing about this:
```
impl Chd {
    #[pyo3(name = "__len__")]
    pub fn len(&self) -> PyResult<usize> {
        Ok(self.inner.header().hunk_count() as usize)
    }
```

**Comment by chyyran** [comment]
ah nvm, if it compiles as `u32` then PyO3 should handle it fine. If you don't need `Result` then you don't need to return `PyResult` either.

**Comment by chyyran** [comment]
lgtm, just run `cargo fmt` and fix the formatting

**Comment by i30817** [comment]
Done. 

It does need a usize btw, just wasn't sure if PyResult is needed for some python protocol involving __len__

**Comment by chyyran** [comment]
released as 0.1.3 on pypi, thanks again!

**Comment by chyyran** [comment]
If you're still interested, I managed to fix the lifetime problem with an `Option` trick to enforce parent lifetimes at runtime in https://github.com/chyyran/chd-rs-py/commit/1c26d412307e9db382f76863a0be9f6bb6fca42f and added a `header` method to `Chd` built off this PR as 0.1.4 on pypi

**Comment by i30817** [comment]
Nice hack. A bit mindbending if you're not used to rust, but i can see how it would be useful in other contexts.

I guess i should stop procrastinating about using metadata to extract track information... it looks so hard though. Especially since the metadata output doesn't look anything like `chdman info`, not that I know how to extract the extents of each track with that anyway.

**Pull request merged**